### ISSUE TITLE
Dockerfile

### ISSUE THREAD

**Issue by ashotmarg** [psathyrella/partis#320]
Hi @psathyrella, thanks for creating "partis".
I've started playing a bit with making a nextflow pipeline for partis. Among other things (ease of use and integration with other pipelines), this would easily solve the potential complications with docker images with batch systems.
I tried to make a docker image using the Dockerfile (running "docker build -t partis_docker ." inside the partis folder) and got an error. 
<img width="1254" alt="Screenshot 2022-12-06 at 15 40 05" src="https://user-images.githubusercontent.com/9378723/205902883-fd239930-f09c-49cd-9838-76ea6a9faa55.png">

The ides was to make sure this step works, so that I could gradually build/include the dependencies (e.g., R and R packages) for simulation, MDS and other analyses.
Any advice on this would be great, not sure if I missed something !
Thanks.

**Comment by psathyrella** [comment]
Great! Sounds like an awesome addition.

I just rebuilt with the same command as you on my machine and it worked fine, so I'm guessing it's something nextflow-specific? I'm not familiar with it at all, but your output looks a bit different so I'm imagining the env is also different. It looks like the build script isn't [correctly cd'ing](https://github.com/psathyrella/partis/blob/dev/bin/build.sh#L14) into the ham dir for you, so I'm guessing the script dir variable isn't getting [set correctly](https://github.com/psathyrella/partis/blob/dev/bin/build.sh#L6)? It looks like I changed it to this weird/complicated way [pretty recently ](https://github.com/psathyrella/partis/commit/2588a9f27a8660d5989b73a6d0dabd00dca27166)so it'd work in the linearham build, but in retrospect it's a bit silly to make it bash dependent. Maybe you can figure out what shell you're in there and suggest something that'll work in both?

![p](https://user-images.githubusercontent.com/2960243/205947014-75fc6bd3-18dd-4956-b927-f34692548530.png)

**Comment by psathyrella** [comment]
ah, ok, then sounds like a difference in the macos vs linux docker versions. If it's possible to suggest a fix that'd be great.

**Comment by KasapovicS** [comment]
When you do just the `git clone` on the repo the partis\packages\ham folder is empty.
This isnt the only one, there are many of them as this repository contains submodules, the entire repo should be ~550MB.
You should instead run

`git clone --recursive https://github.com/psathyrella/partis.git`

This populates all those folders and gets you the code for building the image.

**Comment by psathyrella** [comment]
That's a great point, running the wrong git clone command would cause this error. Do you know why there would be a difference in this (maybe in the COPY behavior) in different docker versions, though?

**Comment by KasapovicS** [comment]
> That's a great point, running the wrong git clone command would cause this error. Do you know why there would be a difference in this (maybe in the COPY behavior) in different docker versions, though?

COPY behavior is the same once docker build starts, that's kinda the point of docker. There are some differences though in the build process on new Macs.

This major difference happens on newer MacOS, due to different processor instructions. On an Intel/AMD Machine (even an older Mac), docker pull will pull the linux AMD64 image with `FROM` instruction or `docker pull`. On a newer Mac using M1/2 processor, docker pull will the pull the linux/arm64/v8 image. This different image might miss some stuff with installations, most notorious one being missing of gcc or the distro not having all python wheels.

But here we quite clearly have a file missing. I might be a bit over confident, because I got the same error as he did - when I cloned it wrong :). Regular clone will give you the ham folder, but it will be empty. Since there was nothing in the ham folder, the Sconstruct file was not found.

**Comment by psathyrella** [comment]
Yes I think you're likely right that this is the issue. It seems absolutely wild to me that this kind of difference exists (like, isn't avoiding this the whole point of docker?). In any case, let me know if there's anything I can do to avoid it, otherwise I guess I can just put a not that it may not work on mac. I've tried setting submodule.recurse in the repo's .git/config, but i don't think there's a way for that to propagate.

**Comment by KasapovicS** [comment]
You could build an image yourself for each version you release and push it somewhere. Like dockerhub.

**Comment by psathyrella** [comment]
Does anyone know if this is still an issue with the new python 3 docker file?

**Comment by psathyrella** [comment]
Oh, and yeah, I used to manually build and push images to docker hub for the year or so between when dockerhub disabled auto builds and I switched to quay. If it would really solve a problem I could potentially do it occasionally, but I'm not sure I've understood how me building partis would get around the problems above? Maybe if I had access to a mac, but I don't -- perhaps a better fix would be adding documentation to help people build their own docker image on mac?
